core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java

```diff
@@ -162,45 +156,40 @@ public void doWork() {
             this.storeUpdateHandler.handleUpdate(messageKey, message);
             try {
               offsetUpdateLock.lock();
-              offsetInSchemasTopic = messageAndMetadata.offset();
+              offsetInSchemasTopic = record.offset();
               offsetReachedThreshold.signalAll();
             } finally {
               offsetUpdateLock.unlock();
             }
           } catch (StoreException se) {
-            log.error("Failed to add record from the Kafka topic" + topic + " the local store");
+            throw new RuntimeException("Failed to add record from the Kafka topic" + topic + " the local store");
           }
         }
       }
-    } catch (ConsumerTimeoutException cte) {
-      // Expect ConsumerTimeout == -1, so this *should* never happen
-      throw new IllegalStateException(
-          "KafkaStoreReaderThread's ConsumerIterator timed out despite expected infinite timeout.");
-    } catch (MessageSizeTooLargeException mstle) {
+    } catch (WakeupException we) {
+      consumer.close();
```

Comment:
It can't just be moved to the finally block, because doWork() is called from a `while(true)` loop. So putting it in `finally` will prematurely close it. However, you're right that there could be a case where the consumer isn't closed. An alternative could be to use an AtomicBoolean, right?